File: Graph/ShortDisInBinMaze/maze.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    int shortestPath(vector<vector<int>>& mat, vector<int>& src, vector<int>& dest) {

        int m = mat.size();
        int n = mat[0].size();

        int sr = src[0];
        int sc = src[1];

        int fr = dest[0];
        int fc = dest[1];

        if (!(0 <= sr && sr < m && 0 <= sc && sc < n))
            return -1;

        if (!(0 <= fr && fr < m && 0 <= fc && fc < n))
            return -1;

        if (!mat[sr][sc] || !mat[fr][fc])
            return -1;

        vector<vector<bool>> visited(m, vector<bool>(n, false));

        queue<pair<pair<int,int>, int>> qu;
        qu.push({{sr, sc}, 0});
        visited[sr][sc] = true;

        int dx[4] = {-1, 0, 1, 0};
        int dy[4] = {0, -1, 0, 1};

        while (!qu.empty()) {
            auto curr = qu.front();
            qu.pop();

            int r = curr.first.first;
            int c = curr.first.second;
            int dis = curr.second;

            if (r == fr && c == fc)
                return dis;

            for (int i = 0; i < 4; i++) {
                int nr = r + dx[i];
                int nc = c + dy[i];

                if (nr >= 0 && nr < m && nc >= 0 && nc < n) {
                    if (!visited[nr][nc] && mat[nr][nc]) {
                        visited[nr][nc] = true;
                        qu.push({{nr, nc}, dis + 1});
                    }
                }
            }
        }

        return -1;
    }
};

File: Graph/ShortDisInBinMaze/maze.py
from collections import deque

class Solution:
    def shortestPath(self, mat: list[list[int]], src: list[int], dest: list[int]) -> int:
        # code here
        m = len(mat)
        n = len(mat[0])

        qu = deque()
        
        sr = src[0]
        sc = src[1]
        
        fr = dest[0]
        fc = dest[1]
        
        if not mat[sr][sc] or not mat[fr][fc]: return -1
        if not (0 <= sr < m and 0 <= sc < n): return -1
        if not (0 <= fr < m and 0 <= fc < n): return -1
        
        visited = [[False for _ in range(n)] for _ in range(m)]
        
        qu.append(((sr, sc),0))
        visited[sr][sc] = True

        dx = [-1, 0, 1, 0]
        dy = [0, -1, 0, 1]

        while qu:
            curr = qu.popleft()
            r = curr[0][0]
            c = curr[0][1]
            dis = curr[1]
            
            if (r,c) == (fr,fc): return dis

            for i in range(4):
                nr = r + dx[i]
                nc = c + dy[i]

                if 0 <= nr < m and 0 <= nc < n:
                    if not visited[nr][nc] and mat[nr][nc]:
                        qu.append(((nr, nc),dis+1))
                        visited[nr][nc] = True

        return -1